### PR TITLE
[FLINK-11871][table-runtime-blink] Introduce LongHybridHashTable to improve performance when join key fits in long

### PR DESCRIPTION
**Pull request by KurtYoung**
## What is the purpose of the change

We can do further optimizations if we know the join key fits in long, a single long field or two integer fields are both ok. For example, we can combine the hash code and actual key into one field, there will be no hash collision, can save a lot of unnecessary logic.

## Brief change log

  - Introduce a new data structure: LongHybridHashTable

## Verifying this change

This change added some unit tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (yes)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

### CODE REVIEW

**Comment by KurtYoung** [comment]
cc @JingsongLi

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/util/SegmentsUtil.java [hunk 1]
```diff
@@ -159,6 +175,47 @@ private static void copyMultiSegmentsToUnsafe(
...
+			MemorySegment[] segments,
+			int offset,
+			int sizeInBytes,
+			DataOutputView target) throws IOException {
+		for (MemorySegment sourceSegment : segments) {
+			int curSegRemain = sourceSegment.size() - offset;
+			if (curSegRemain > 0) {
+				int copySize = Math.min(curSegRemain, sizeInBytes);
+
+				byte[] bytes = new byte[copySize];
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/util/SegmentsUtil.java]: get bytes from allocateReuseBytes or add TODO?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/hashtable/LongHashPartition.java [hunk 2]
```diff
@@ -0,0 +1,947 @@
...
+	}
+
+	void releaseBuckets() {
+		if (buckets != null) {
+			longTable.returnAll(Arrays.asList(buckets));
+			buckets = null;
+		}
+	}
+
+//	public void append(long key, BinaryRow row) throws IOException {
```
Comment (JingsongLi) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/hashtable/LongHashPartition.java]: delete it?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/util/SegmentsUtil.java (hunk 1 above)
Comment (KurtYoung) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/util/SegmentsUtil.java]: i think allocate reuse bytes make sense

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/hashtable/LongHashPartition.java (hunk 2 above)
Comment (KurtYoung) [reply flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/runtime/hashtable/LongHashPartition.java]: ok

**Comment by KurtYoung** [comment]
@JingsongLi Thanks for the reviewing, i will address the comments and merge this later.

**Pull request closed without merging**